pkg/lint/rules/template.go

```diff
@@ -136,6 +139,32 @@ func Templates(linter *support.Linter, values map[string]interface{}, namespace
 	}
 }
 
+// validateTopIndentLevel checks that the content does not start with an indent level > 0.
+//
+// This error can occur when a template accidentally inserts space. It can cause
+// unpredictable errors dependening on whether the text is normalized before being passed
+// into the YAML parser. So we trap it here.
+//
+// See https://github.com/helm/helm/issues/8467
+func validateTopIndentLevel(content string) error {
+	// Read lines until we get to a non-empty one
+	scanner := bufio.NewScanner(bytes.NewBufferString(content))
+	for scanner.Scan() {
+		line := scanner.Text()
+		// If line is empty, skip
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+		// If it starts with one or more spaces, this is an error
+		if strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t") {
```

Comment:
I've been digging through YAML parsers, validators, and the specification. I've learned a few things...

If everything is indented the same level, it's valid. So, if the first line is indented two spaces but the others aren't it's invalid. But, if each line has that two space indentation it's valid. The YAML parser we use handles the all the things are indented case. So the two following cases are valid:

```
foo: bar
baz: quz
```
and
```
  foo: bar
  baz: quz
```
but, the following is invalid...
```
  foo: bar
baz: quz
```
Different tools handle the failure case differently. I don't see, in the spec, a correct way to handle this. Some tools take the lines with the indentation and parse up to the point where the outdent happens and then ignore the rest. The YAML parser we have appears to do that.

I'm not entirely sure of the right way to check this prefix situation.

Reply:
Linter's job is mainly to be prescriptive. Indenting "at the top level" is a bad practice, so linter should warn about it.